src/core/shutdown.c

```diff
@@ -159,6 +163,78 @@ static int switch_root_initramfs(void) {
         return switch_root("/run/initramfs", "/oldroot", false, MS_BIND);
 }
 
+static bool sync_making_progress(unsigned long long *prev_dirty) {
+        _cleanup_free_ char *dirty_string = NULL;
+        unsigned long long dirty = 0;
+        int r;
+
+        r = get_proc_field("/proc/meminfo", "Dirty", WHITESPACE, &dirty_string);
+        if (r < 0) {
+                log_error_errno(r, "Failed to retrieve Dirty from /proc/meminfo: %m");
+                return false;
+        }
+
+        r = safe_atollu(dirty_string, &dirty);
+        if (r < 0) {
+                log_error_errno(r, "Failed to parse Dirty from /proc/meminfo: %s: %m",
+                                dirty_string);
+                return false;
+        }
+
+        if (*prev_dirty > dirty)
+                r = true;
+        else
+                r = false;
+
+        *prev_dirty = dirty;
+
+        return r;
+}
+
+static void sync_with_progress(void) {
+        unsigned checks;
+        pid_t pid;
+        int r;
+        unsigned long long dirty = -1;
+
+        BLOCK_SIGNALS(SIGCHLD);
+
+        /* Due to the possiblity of the sync operation hanging, we fork
+         * a child process and monitor the progress. If the timeout
+         * lapses, the assumption is that that particular sync stalled. */
+        pid = fork();
+        if (pid < 0)
+                log_error_errno(errno, "Failed to fork: %m");
+
+        if (pid == 0) {
+                /* Start the sync operation here in the child */
+                log_info("Syncing filesystems and block devices");
```

Comment:
why print this in the child? parent appears like the better place, otherwise the order is theoretically undefined in which this appears, as we log other stuff later on from the parent too. Just move this after this if block...